apps/full-site-editing/full-site-editing-plugin/components/post-autocomplete/index.js

```diff
@@ -0,0 +1,121 @@
+/* eslint-disable wpcalypso/jsx-classname-namespace */
+/**
+ * External dependencies
+ */
+import { map, throttle } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import apiFetch from '@wordpress/api-fetch';
+import { Button, Popover, Spinner, TextControl } from '@wordpress/components';
+import { withState } from '@wordpress/compose';
+import { select } from '@wordpress/data';
+import { __ } from '@wordpress/i18n';
+import { addQueryArgs } from '@wordpress/url';
+
+/**
+ * Internal dependencies
+ */
+import './style.scss';
+
+const updateSuggestions = throttle( async ( search, setState ) => {
+	setState( {
+		loading: true,
+		showSuggestions: true,
+		suggestions: [],
+	} );
+
+	const suggestions = await apiFetch( {
+		path: addQueryArgs( '/wp/v2/search', {
+			context: 'embed',
+			per_page: 20,
+			search,
+		} ),
+	} );
```

Comment:
we probably want to make sure this `throttle( apiFetch )` combination works the way we expect it to

the `fetchAllMiddleware()` might do some funky business here and we could potentially end up loading all the posts on the first few letters of autocomplete then proceed to reload a subset of those on and on until the end

we also probably want a `debounce()` instead of a `throttle` because we want the last-typed letters to trigger the last autocomplete search and `throttle()` might drop them if they were typed quickly after the previous query

Reply:
You're correct, I've meant to use `debounce` and I've also forgot to add the wait time.
It looks to me it's working correctly, I can't find a case in which this would end up in an infinite loop of some sort. 🤔 

(Also, the search is only triggered when the query is 2 or more characters long, and returns 20 results top)

Reply:
> returns 20 results top

this is strange to me. I can't figure out why it's the case but I can't get my search results to fetch all of the result pages. I see a proper `Link` header when not making calls to dotcom sites and the botched link header _on_ dotcom sites. I thought that the way we mangle this header in our custom middleware replacements could be behind the behavior but I had trouble getting the auto-paging to work in a core-only test.

here's what I'm worried about: I think that it's a bug somewhere preventing the `fetchAllMiddleware` from continuing to pull in all successive search results, so even with `per_page=20` it _should_ fetch all the results automatically. That's my understanding, at least. I am not seeing the middleware do that but I'm at a loss for an explanation of why.

Reply:
I'm not familiar with `api-fetch`, but I gave a look at `fetchAllMiddleware` and as it turns out it's only applied when the request explicitly asks for all results by "unbounding" the limit with `per_page=-1`:

https://github.com/WordPress/gutenberg/blob/774713c0467779a9651983ec96609a6637811132/packages/api-fetch/src/middlewares/fetch-all-middleware.js#L47-L50

https://github.com/WordPress/gutenberg/blob/774713c0467779a9651983ec96609a6637811132/packages/api-fetch/src/middlewares/fetch-all-middleware.js#L33-L37